src/java/org/apache/cassandra/index/sai/utils/IndexTermType.java

```diff
@@ -0,0 +1,858 @@
+                        case KEYS:
+                            return cell.path().get(0);
+                        case VALUES:
+                            return cell.buffer();
+                        case KEYS_AND_VALUES:
+                            return CompositeType.build(ByteBufferAccessor.instance, cell.path().get(0), cell.buffer());
+                    }
+            }
+        }
+        return cell.buffer();
+    }
+
+    private AbstractType<?> collectionCellValueType(AbstractType<?> type, IndexTarget.Type indexType)
+    {
+        CollectionType<?> collection = ((CollectionType<?>) type);
+        switch (collection.kind)
+        {
+            case LIST:
+                return collection.valueComparator();
+            case SET:
+                return collection.nameComparator();
+            case MAP:
+                switch (indexType)
+                {
+                    case KEYS:
+                        return collection.nameComparator();
+                    case VALUES:
+                        return collection.valueComparator();
+                    case KEYS_AND_VALUES:
+                        return CompositeType.getInstance(collection.nameComparator(), collection.valueComparator());
+                }
+            default:
+                throw new IllegalArgumentException("Unsupported collection type: " + collection.kind);
+        }
+    }
+
+    private boolean isCompositePartition()
+    {
+        return hasProperty(IS_COMPOSITE_PARTITION);
+    }
+
+    /**
+     * Returns <code>true</code> if given {@link AbstractType} is {@link InetAddressType}
+     */
+    private boolean isInetAddress()
+    {
+        return hasProperty(IS_INET_ADDRESS);
+    }
+
+    /**
+     * Compares 2 InetAddress terms by ensuring that both addresses are represented as
+     * ipv6 addresses.
+     */
+    private int compareInet(ByteBuffer b1, ByteBuffer b2)
+    {
+        assert isIPv6(b1) && isIPv6(b2);
+
+        return FastByteOperations.compareUnsigned(b1, b2);
+    }
+
+    private boolean isIPv6(ByteBuffer address)
+    {
+        return address.remaining() == INET_ADDRESS_SIZE;
+    }
+
+    /**
+     * Returns <code>true</code> if given {@link AbstractType} is {@link IntegerType}
+     */
+    private boolean isBigInteger()
+    {
+        return hasProperty(IS_BIG_INTEGER);
+    }
+
+    /**
+     * Returns <code>true</code> if given {@link AbstractType} is {@link DecimalType}
+     */
+    private boolean isBigDecimal()
+    {
+        return hasProperty(IS_BIG_DECIMAL);
+    }
+
+    private boolean isLong()
+    {
+        return hasProperty(IS_LONG);
+    }
+
+    /**
+     * Encode a {@link InetAddress} into a fixed width 16 byte encoded value.
+     * <p>
+     * The encoded value is byte comparable and prefix compressible.
+     * <p>
+     * The encoding is done by converting ipv4 addresses to their ipv6 equivalent.
+     */
+    private ByteBuffer encodeInetAddress(ByteBuffer value)
```

Comment:
nit: Looks like this, `encodeBigInteger()`, and `encodeDecimal()` can still be `static`?